src/server/protocol_client.cc

```diff
@@ -0,0 +1,333 @@
+// Copyright 2022, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+#include "server/protocol_client.h"
+
+extern "C" {
+#include "redis/rdb.h"
+}
+
+#include <absl/cleanup/cleanup.h>
+#include <absl/flags/flag.h>
+#include <absl/functional/bind_front.h>
+#include <absl/strings/escaping.h>
+#include <absl/strings/str_cat.h>
+#include <absl/strings/strip.h>
+
+#include <boost/asio/ip/tcp.hpp>
+
+#include "base/logging.h"
+#include "facade/dragonfly_connection.h"
+#include "facade/redis_parser.h"
+#include "server/error.h"
+#include "server/journal/executor.h"
+#include "server/journal/serializer.h"
+#include "server/main_service.h"
+#include "server/rdb_load.h"
+#include "strings/human_readable.h"
+
+ABSL_FLAG(std::string, masterauth, "", "password for authentication with master");
+
+#define RETURN_ON_BAD_RESPONSE(x)                                                               \
+  do {                                                                                          \
+    if (!(x)) {                                                                                 \
+      LOG(ERROR) << "Bad response to \"" << last_cmd_ << "\": \"" << absl::CEscape(last_resp_); \
+      return std::make_error_code(errc::bad_message);                                           \
+    }                                                                                           \
+  } while (false)
+
+namespace dfly {
+
+using namespace std;
+using namespace util;
+using namespace boost::asio;
+using namespace facade;
+using absl::GetFlag;
+using absl::StrCat;
+
+namespace {
+
+int ResolveDns(std::string_view host, char* dest) {
```

Comment:
Do you think this might be a good chance to use Helio's resolve DNS?
(it's better because it doesn't use blocking APIs, and also saves this messy function)

Reply:
We should do it, but not in this PR :smile: